compiler/rustc_trait_selection/src/traits/mod.rs

```diff
@@ -454,6 +456,144 @@ fn subst_and_check_impossible_predicates<'tcx>(
     result
 }
 
+#[derive(Clone, Debug)]
+enum VtblSegment<'tcx> {
+    MetadataDSA,
+    TraitOwnEntries { trait_ref: ty::PolyTraitRef<'tcx>, emit_vptr: bool },
+}
+
+/// Retrieve a generator that indicates the internal arrangements of vtable.
+fn vtable_planner<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    trait_ref: ty::PolyTraitRef<'tcx>,
+) -> impl Generator<(), Yield = VtblSegment<'tcx>> + 'tcx {
```

Comment:
Instead of a generator can you use a callback?

Reply:
Good idea! Thanks.

Reply:
Done.